### ISSUE TITLE
Claude Code Review Report(with multiple, simple to fix bugs)

### ISSUE THREAD

**Issue by qarmin** [slint-ui/slint#12175]
Recently, I ran Claude Code across several projects(like image-rs, lofty of symphonia) to look for potential issues(logic bugs, copy paste errors, architecture problems, panics etc.).

In this projects, a significant portion of reported findings turned out to be incorrect or minor false positives, but a non-trivial subset was valid and included real issues ranging from incorrect comments to actual logic bugs. As a result, the findings generally require manual validation to separate noise from actionable problems.

Full Report - [slint_20260620.html](https://github.com/user-attachments/files/29161785/slint_20260620.html)

I already fixed some (very easy) issues, but there are still many more that could be addressed.

Example findings (this is only a subset of the findings, specifically those most likely to be actual bugs; for the full list, see the report above):

=====================================

### CPY_1 `HIGH` ⚡ easy

**Description:** Bounce easing curves are mapped to the wrong C++ enum tags. `EaseOutBounce` is generated as `EasingCurve::Tag::EaseOutElastic` and `EaseInOutBounce` as `EaseInOutElastic`. These are clearly copy-paste errors from the elastic cases just above; a `.slint` file using `ease-out-bounce` / `ease-in-out-bounce` animations will silently get elastic easing in the C++ backend instead of bounce.

**Locations:**

```
internal/compiler/generator/cpp.rs:4244-4249
 4244 |         Expression::EasingCurve(EasingCurve::EaseOutBounce) => {
 4245 |             "slint::cbindgen_private::EasingCurve::Tag::EaseOutElastic".into()
 4246 |         }
 4247 |         Expression::EasingCurve(EasingCurve::EaseInOutBounce) => {
 4248 |             "slint::cbindgen_private::EasingCurve::Tag::EaseInOutElastic".into()
 4249 |         }
```

**Fix:** Emit `EasingCurve::Tag::EaseOutBounce` and `EasingCurve::Tag::EaseInOutBounce` respectively for those two arms.

### CPY_4 `HIGH` ⚡ easy

**Description:** check_popup_layout has a copy-paste bug: the second `if` checks `fixed_height` again instead of `fixed_width`. As a result, a popup window with a fixed width (but non-fixed height) never has its width property frozen via adjust_window_layout(&p.component, "width"), so the runtime can still overwrite its width; and a fixed-height popup runs the width adjustment keyed off the height flag. Compare with the correct check_window_layout directly above (lines 2192-2199) which checks fixed_height then fixed_width.

**Locations:**

```
internal/compiler/passes/lower_layout.rs:2201-2211
 2201 | pub fn check_popup_layout(component: &Rc<Component>) {
 2202 |     component.popup_windows.borrow().iter().for_each(|p| {
 2203 |         if p.component.root_constraints.borrow().fixed_height {
 2204 |             adjust_window_layout(&p.component, "height");
 2205 |         }
 2206 | 
 2207 |         if p.component.root_constraints.borrow().fixed_height {
 2208 |             adjust_window_layout(&p.component, "width");
 2209 |         }
 2210 |     });
 2211 | }
```

**Fix:** Change line 2207 from `if p.component.root_constraints.borrow().fixed_height {` to `if p.component.root_constraints.borrow().fixed_width {`.

### CPY_5 `HIGH` ⚡ easy

**Description:** The color palette built in the gallery has several wrong RGB values, clearly the
result of copy-paste edits that were not finished:
- "maroon" is defined as rgb(128, 0, 255) - that is a purple/violet, not maroon
  (should be rgb(128, 0, 0)).
- "red" is defined as rgb(0, 255, 0) - that is pure green, identical to "lime",
  not red (should be rgb(255, 0, 0)).
- "sliver" is a misspelling of "silver".
So selecting "red" or "maroon" in the gallery shows the wrong color swatch.

**Locations:**

```
ui-libraries/material/examples/gallery/src/lib.rs:41-58
   41 |     let colors = VecModel::from_slice(&[
   42 |         color_item("aqua", 0, 255, 255, ui),
   43 |         color_item("black", 0, 0, 0, ui),
   44 |         color_item("blue", 0, 0, 255, ui),
   45 |         color_item("fuchsia", 255, 0, 255, ui),
   46 |         color_item("gray", 128, 128, 128, ui),
   47 |         color_item("green", 0, 128, 0, ui),
   48 |         color_item("lime", 0, 255, 0, ui),
   49 |         color_item("maroon", 128, 0, 255, ui),
   50 |         color_item("navy", 0, 0, 128, ui),
   51 |         color_item("olive", 128, 128, 0, ui),
   52 |         color_item("purple", 128, 0, 128, ui),
   53 |         color_item("red", 0, 255, 0, ui),
   54 |         color_item("sliver", 192, 192, 192, ui),
   55 |         color_item("teal", 0, 128, 128, ui),
   56 |         color_item("white", 255, 255, 255, ui),
   57 |         color_item("yellow", 255, 255, 0, ui),
   58 |     ]);
```

**Fix:** Correct the values:
    color_item("maroon", 128, 0, 0, ui),
    ...
    color_item("red", 255, 0, 0, ui),
    color_item("silver", 192, 192, 192, ui),

### CPY_9 `LOW` ⚡ easy

**Description:** The hard-coded year model in DatePickerPopup is missing 2030. The list jumps
2028, 2029, 2031, 2032 ... so 2030 cannot be selected from the year picker. This
looks like a hand-typed list with a dropped entry.

**Locations:**

```
ui-libraries/material/src/ui/components/date_picker.slinat:354
  354 |     property <[int]> year_model: [2024, 2025, 2026, 2027, 2028, 2029, 2031, 2032, 2033, 2034, 2035, 2036, 2037, 2038, 2039, 2040, 2041, 2042, 2043];
```

**Fix:** Insert 2030 between 2029 and 2031 in the year_model list.

### LOGIC_1 `HIGH` ⚡ easy

**Description:** `operator!=` for `Slice<T>` is defined in terms of itself: `return !(a != b);`. This is
unconditional infinite recursion (no base case), so any use of `slice_a != slice_b` will
either stack-overflow at runtime or, more likely, be miscompiled by the optimizer into a
trap/garbage result. It should delegate to `operator==`, i.e. `return !(a == b);`. The
`operator==` directly above it is correct, which makes the typo obvious.

**Locations:**

```
api/cpp/include/private/slint_sharedvector.h:242-246
  242 | template<typename T>
  243 | bool operator!=(cbindgen_private::Slice<T> a, cbindgen_private::Slice<T> b)
  244 | {
  245 |     return !(a != b);
  246 | }
```

**Fix:** Replace the body with `return !(a == b);`.

### LOGIC_2 `HIGH`

**Description:** `SoftwareRenderer::render_by_line` only handles `Rgb565Pixel` and `Rgb8Pixel`, but its
fall-through `static_assert` message and condition reference `Rgba8Pixel`:
`static_assert(std::is_same_v<PixelType, Rgba8Pixel> || std::is_same_v<PixelType, Rgb565Pixel>, ...)`.
`Rgba8Pixel` is NOT one of the supported branches above (only Rgb565Pixel/Rgb8Pixel are), so
the assert condition is wrong: instantiating with `Rgba8Pixel` would silently fall into the
`else` branch and the `static_assert` would actually PASS (because the condition is true for
Rgba8Pixel), leaving the function returning an uninitialized/garbage `PhysicalRegion` with no
FFI call made. The intended supported types are Rgb8Pixel and Rgb565Pixel.

**Locations:**

```
api/cpp/include/slint-platform.h:815-826
  815 |         if constexpr (std::is_same_v<PixelType, Rgb565Pixel>) {
  816 |             return PhysicalRegion { cbindgen_private::slint_software_renderer_render_by_line_rgb565(
  817 |                     inner, process_line_fn, &process_line_callback) };
  818 |         } else if constexpr (std::is_same_v<PixelType, Rgb8Pixel>) {
  819 |             return PhysicalRegion { cbindgen_private::slint_software_renderer_render_by_line_rgb8(
  820 |                     inner, process_line_fn, &process_line_callback) };
  821 |         } else {
  822 |             static_assert(std::is_same_v<PixelType, Rgba8Pixel>
  823 |                                   || std::is_same_v<PixelType, Rgb565Pixel>,
  824 |                           "Unsupported PixelType. It must be either Rgba8Pixel or Rgb565Pixel");
  825 |         }
  826 |     }
```

**Fix:** Change the else-branch assert to reference the actually-supported types, e.g.
`static_assert(std::is_same_v<PixelType, Rgb8Pixel> || std::is_same_v<PixelType, Rgb565Pixel>,
"Unsupported PixelType. It must be either Rgb8Pixel or Rgb565Pixel");`

### LOGIC_12 `HIGH`

**Description:** Always-true condition in the time picker clock. The selection-line `Path` and
the picker dot are gated by `if root.current_index >= 0 || root.current_index <
root.model.length`. Because it is an OR of two complementary half-open ranges,
the expression is true for every integer value of `current_index` (including the
sentinel -1 used elsewhere in this file). The intent was clearly `>= 0 &&
< model.length`. As written the guard never hides the elements, so the clock
hand is drawn even when there is no valid selection.

**Locations:**

```
ui-libraries/material/src/ui/components/time_picker.slint:71
   71 |         if root.current_index >= 0 || root.current_index < root.model.length: Path {
```

**Fix:** Use `if root.current_index >= 0 && root.current_index < root.model.length`.

### LOGIC_21 `MEDIUM` ⚡ easy

**Description:** ComboBox-popup list item render clobbers the accumulated style state. Line 158 ORs `initial_state` (which carries `State_Active` when the window is active) into `option.state`, but line 161 then does a plain assignment `option.state = QStyle::State_Enabled;`, wiping out the `initial_state` contribution. So combobox popup items are always drawn as if the window were inactive, defeating the whole purpose of computing `initial_state`. The non-combobox branch correctly uses `|=` throughout.

**Locations:**

```
internal/backends/qt/qt_widgets/listviewitem.rs:158-161
  158 |                 option.state |= QStyle::State(initial_state);
  159 |                 option.rect = QRect(QPoint(), size / dpr);
  160 |                 option.menuRect = QRect(QPoint(), size / dpr);
  161 |                 option.state = QStyle::State_Enabled;
```

**Fix:** Change line 161 to `option.state |= QStyle::State_Enabled;` (and drop the now-redundant line 158).

### LOGIC_64 `MEDIUM` ⚡ easy

**Description:** The format CLI never formats `.60` files. The match arm compares the extension
against the string `".60"` (with a leading dot), but `Path::extension()` returns
the extension *without* the dot (e.g. `"60"`). So a `.60` file always falls
through to the catch-all arm and is emitted verbatim instead of being formatted.
The accompanying comment ("Formatting .60 files because of backwards
compatibility") documents the opposite of what actually happens.

**Locations:**

```
tools/lsp/fmt/tool.rs:111
  111 |         Some(ext) if ext == "slint" || ext == ".60" => process_slint_file(source, path, file),
```

**Fix:** Compare against the un-dotted extension:
`Some(ext) if ext == "slint" || ext == "60" => process_slint_file(source, path, file),`
